models/gpr.py

```diff
@@ -0,0 +1,84 @@
+from abc import ABC, abstractmethod
+from sklearn.metrics import mean_squared_error
+from sklearn.gaussian_process.kernels import RBF, ConstantKernel, Matern
+from sklearn.gaussian_process import GaussianProcessRegressor
+import pandas as pd
+import numpy as np
+import matplotlib as mpl 
+from matplotlib import pyplot as plt
+from models.model import Model
+class GPR(Model):
+	def __init__(self, dist_array, user_name, dep_name, length, file, lat_n, lon_n):
+		self.dist = dist_array#array-like
+		self.user = file[user_name].values#array-like
+		self.dep = file[dep_name].values#array-like
+		self.depth = []
+		for i in range(len(self.dep)):
+			self.depth.append(self.dep[i] * -1)
+		self.leng = length#int
+		self.gp = GaussianProcessRegressor(normalize_y=True, kernel=Matern()+ConstantKernel(), alpha = 0.0001)
+		self.x_coor = np.linspace(min(self.dist), max(self.dist), self.leng)
+		self.y_coor = np.linspace(min(self.depth), max(self.depth), self.leng)
+		self.xx, self.yy = np.meshgrid(self.x_coor, self.y_coor)
+		self.x = []
+		self.y = []
+		self.xy = []
+		for i in range(len(self.xx)):
+			for j in range(len(self.xx[i])):
+				self.xy.append([self.xx[i][j], self.yy[i][j]])
+				self.x.append(self.xx[i][j])
+				self.y.append(self.yy[i][j])
+		self.x_arr = []
+		for i in range(len(self.depth)):
+			self.x_arr.append([self.dist[i], self.depth[i]])
+		self.gp = self.gp.fit(self.x_arr, self.user)
```

Comment:
Yeah so I wasn't sure if Python was like Java in the sense that you can use defined methods in other methods within the same class, so I just fit in the constructor. It also works out because if I fit in the fit() method, then I would have to fit more than once for the other methods in the class, which need a fit model. What I think I'm going to do is put an if-else statement in the constructor that checks if the model is supposed to be 2d or 3d, and then just make one model so that I only have to fit once the whole time, which is the least expensive option.

Reply:
I think I see what you're talking about. I think a better option would still be to not fit in the object constructor. Instead, you can set up your gaussian process model like this:
```
def __init__(self, ...):
    self.gp_model = GaussianProcessRegressor(...)
    self.gp_model_is_fitted = False
...
@property
def gp(self):
    if not self.gp_model_is_fitted:
        self.gp_model.fit()
    return self.gp_model
...
def some_other_method(self):
  self.gp.predict(...)  # will only fit self.gp_model once
```
The @property tag will allow the method below to be called as if it was a class attribute (ie self.gp instead of self.gp()).

This still ensures you only fit once, but it fits in a "lazy" manner, meaning it only runs the heavy computation when it really needs to. It's not super important that we do this, but it could be a good exercise for you to see how this works.

Reply:
Made this change